record-and-playback/presentation/playback/presentation/0.9.0/lib/writing.js

```diff
@@ -386,64 +377,94 @@ function runPopcorn() {
+    } else {
+      return true;
+    }
+  }
+
+  var drawShape = function(shapes, shapeElement, shapeTime, currentTime) {
+    var visibleShape = false;
+    if (shapeTime < currentTime) {
+      // Check if it is a main shape
+      if(main_shapes_ids.indexOf(shapeElement.getAttribute("id")) > -1) {
+        var undoTime = parseFloat(shapeElement.getAttribute("undo"));
+        // And hasn't been undone
+        if(undoTime === -1 || undoTime > currentTime) {
+          shapeElement.style.visibility = "visible";
+          visibleShape = true;
+        }
+      }
+    } else if (shapeTime == currentTime) {
+      // Only makes visible the last drawing of a given shape
+      var shape = shapeElement.getAttribute("shape");
+      var shapeIndex = shapes.indexOf(shape);
+      if (shapeIndex > -1) {
+        shapes.splice(shapeIndex, 1);
+        shapeIndex = shapes.indexOf(shape);
+        if (shapeIndex === -1) {
+          shapeElement.style.visibility = "visible";
+          visibleShape = true;
+        }
+      }
+    }
+    // It's not a visible shape, force it to be hidden
+    if (!visibleShape) {
+      shapeElement.style.visibility = "hidden";
+    }
+  }
+
+  var lastShapeTime = -1;
+  var updateShapes = function(time) {
+    var svgDocument = null;
+    var shapes = get_shapes_in_time(time);
+    var currentTime = parseFloat(time);
+
+    if (svgobj.contentDocument) svgDocument = svgobj.contentDocument;
+    else svgDocument = svgobj.getSVGDocument('svgfile');
+
+    // Abort if there is no SVG available
+    if (svgDocument == null) return;
+
+    if (regularPlaybackFlow(currentTime, lastShapeTime)) {
+      for (var i = 0; i < shapesArray.length; i++) {
+        var shapeTime = parseFloat(shapesArray[i].getAttribute("timestamp"));
+        if (shapeTime <= currentTime) {
+          // Draw only inside this interval
+          var shapeElement = svgDocument.getElementById(shapesArray[i].getAttribute("id"));
+          drawShape(shapes, shapeElement, shapeTime, currentTime);
+        } else {
+          break;
+        }
+      }
+    } else {
+      for (var i = 0; i < shapesArray.length; i++) {
+        var shapeTime = parseFloat(shapesArray[i].getAttribute("timestamp"));
+        var shapeElement = svgDocument.getElementById(shapesArray[i].getAttribute("id"));
+        drawShape(shapes, shapeElement, shapeTime, currentTime);
+      }
+    }
+    lastShapeTime = currentTime;
+  }
+
   var p = new Popcorn("#video");
   //update 60x / second the position of the next value.
   p.code({
       start: 1, // start time
       end: p.duration(),
       onFrame: function(options) {
-        //console.log("**Popcorn video onframe");
-        if(!((p.paused() === true) && (p.seeking() === false))) {
+        if ( (!p.paused() || p.seeking()) && (p.currentTime() - lastFrameTime >= 0.1) ) {
+          lastFrameTime = p.currentTime();
           var t = p.currentTime().toFixed(1); //get the time and round to 1 decimal place
```

Comment:
There are three calls to p.currentTime(), wouldn't be better to store it in a variable?
Also, we probably should do Math.abs(p.currentTime() - lastFrameTime) >= 0.1